README
README for udevil v0.4.4


Homepage:
    http://ignorantguru.github.io/udevil/

Downloads:
    https://github.com/IgnorantGuru/udevil/tree/pkg

Official Forum (for udevil & SpaceFM):
    https://sourceforge.net/p/spacefm/discussion/general/


CONTENTS

    DESCRIPTION
    PACKAGES
    INSTALLER
    BUILD
    BUILD NEXT
    CREATE DEB PACKAGE
    POST INSTALL
        Kernel Polling
        Networks and Files
        NTFS-3G
        Set SUID
        systemd Service
    ENABLE KERNEL POLLING
    UNINSTALL
    TRANSLATE
    LICENSE


DESCRIPTION

    Mounts and unmounts removable devices and networks without a password (set
    suid), shows device info, monitors device changes.  Emulates mount's and
    udisks's command line usage and udisks v1's output.  Includes the devmon
    automounting daemon.

    Run 'udevil --help' for usage.  See /etc/udevil/udevil.conf for config.
        http://ignorantguru.github.com/udevil/udevil--help.html
        http://ignorantguru.github.com/udevil/udevil.conf.txt


PACKAGES

    Find or add packages, forum threads, and other info for your distro:
        https://github.com/IgnorantGuru/spacefm/wiki/Distros

    For other distros:
        See INSTALLER below

    NOTE: See POST INSTALL section below for post-installation steps.


INSTALLER

    A self-extracting installer is available (udevil-x.x.x-installer.sh):
        https://github.com/IgnorantGuru/udevil/tree/pkg
            (To save a file, click on its filename and click 'View Raw')

    The installer requires the build dependencies listed below in BUILD.  If
    any dependencies are missing, examine the error, install missing packages,
    and try again.

    The installer MUST be run in a terminal.  It automatically builds and
    installs.  Run it like this:

        bash udevil-x.x.x-installer.sh

    or to see options:

        bash udevil-x.x.x-installer.sh --help

    To reinstall or upgrade, just run the latest installer again.

    NOTE: See POST INSTALL section below for post-installation steps.


BUILD

    1)  Install required build dependencies (below are Debian package names -
        packages names on your distro may vary but should be similar):

            libglib2.0-0 libglib2.0-dev libudev0 (>=143) libudev-dev
            autotools-dev build-essential fakeroot intltool pkg-config

        Also recommended for network support: cifs-utils curlftpfs sshfs

        Also recommended if using devmon: eject spacefm|zenity

        configure will report anything missing when you run it as shown below.

    2)  Use these commands to download the udevil tarball and build:

        # Download & Extract
        mkdir /tmp/udevil-build && cd /tmp/udevil-build
        wget -O udevil.tar.gz https://github.com/IgnorantGuru/udevil/tarball/master
        tar xzf udevil.tar.gz
        cd IgnorantGuru-udevil-*

        # Build & Install
        ./configure --prefix=/usr
        make
        sudo make install

        # Remove Temporary Files
        cd / && rm -rf /tmp/udevil-build

    NOTES:

        * If configure reports missing dependencies, install them and run
          configure again before proceeding to make.

        * By default, systemd configuration files for devmon are installed to:
              /etc/conf.d/devmon
              /usr/lib/systemd/system/devmon@.service
          To skip installation of these files, use configure option
          --disable-systemd   In either case, systemd is NOT required.

        * See POST INSTALL section below for post-installation steps.


BUILD NEXT

    The git 'next' branch of udevil is a work in progress which eventually
    becomes the next release version.  To build from this branch, install the
    required build dependencies listed above in BUILD, then:

    # Download & Extract
    mkdir /tmp/udevil-next && cd /tmp/udevil-next
    wget -O udevil-next.tar.gz https://github.com/IgnorantGuru/udevil/tarball/next
    tar xzf udevil-next.tar.gz
    cd IgnorantGuru-udevil-*

    # Build, & Install
    ./configure --prefix=/usr
    make
    sudo make install

    # Remove Temporary Files
    cd / && rm -rf /tmp/udevil-next


CREATE DEB PACKAGE

    To build a deb file of udevil on your Debian-based distro:

    1)  Install build dependencies (see BUILD section above) plus 'debhelper'.

    2)  Use these commands to download and extract the udevil tarball (be sure
        to USE A TMP BUILD PATH THAT DOES NOT CONTAIN SPACES):

        mkdir /tmp/udevil-build && cd /tmp/udevil-build  # no spaces in path
        # Note: you can change "master" to "next" if you want the next branch:
        wget -O udevil.tar.gz https://github.com/IgnorantGuru/udevil/tarball/master
        tar xzf udevil.tar.gz
        cd IgnorantGuru-udevil-*

    3)  Move the distros/debian directory into the source directory:

        mv distros/debian .

    4)  Build package:  (the package files will appear in the parent directory)

        dpkg-buildpackage -b -uc
        cd .. && rm -rf IgnorantGuru-udevil-* udevil.tar.gz

    5)  To install the package you built:

        dpkg -i udevil*.deb


POST INSTALL

    Kernel Polling
    --------------

    You may need to enable kernel polling for device media to be detected by
    udevil.  See ENABLE KERNEL POLLING section below.


    Networks and Files
    ------------------

    By default, /etc/udevil/udevil.conf is set to allow only local fileystems
    and ISO files to be mounted, with mounting of networks disallowed.  To
    allow networks and files to be mounted, in /etc/udevil/udevil.conf set:

        allowed_types = $KNOWN_FILESYSTEMS, file, cifs, nfs, curlftpfs, sshfs, davfs

    You may also need to install curlftpfs or ftpfs (ftp://), cifs or smbfs
    (smb://), sshfs (ssh://), and davfs2 (http://).


    NTFS-3G
    -------

    If local ntfs filesystems aren't mounted writable, you may need to
    configure your system to mount ntfs with ntfs-3g (on some distros this is
    already done).  For example:

        sudo ln -s /sbin/mount.ntfs-3g /sbin/mount.ntfs

        # OR if mount.ntfs-3g is located in /usr/bin:
        sudo ln -s /usr/bin/mount.ntfs-3g /usr/bin/mount.ntfs


    Set SUID
    --------

    After installing udevil, /usr/bin/udevil should have the suid bit already
    set.  If not, set it like this:

        sudo chown root:root /usr/bin/udevil
        sudo chmod u+s,go-s,ugo+x /usr/bin/udevil
        ls -l /usr/bin/udevil
            -rwsr-xr-x 1 root root 226625 May 22 08:13 /usr/bin/udevil

    OR, to restrict execution of udevil to the 'plugdev' group only:

        sudo chown root:plugdev /usr/bin/udevil
        sudo chmod u+s,go-s,o-x /usr/bin/udevil
        ls -l /usr/bin/udevil
            -rwsr-xr-- 1 root plugdev 226625 May 22 08:13 /usr/bin/udevil

    OR, if you don't want to use udevil for mounting, you can unset suid:
        sudo chown root:root /usr/bin/udevil
        sudo chmod ugo-s,ugo+x /usr/bin/udevil
        ls -l /usr/bin/udevil
            -rwxr-xr-x 1 root root 226625 May 22 08:13 /usr/bin/udevil

    You can also limit users and groups by editing /etc/udevil/udevil.conf


    systemd Service
    ---------------

    To start devmon as a systemd service:
        systemctl start devmon@<user>


ENABLE KERNEL POLLING

    You may need to enable kernel polling for device media to be detected by
    udevil.  For example, if you insert a CD and udevil's monitor command
    doesn't report a change to the device, or udevil doesn't see a filesystem
    on the device, this is a symptom that kernel polling is not enabled.

    Kernel polling is a newer feature of the Linux kernel and udev, so some
    distros don't yet have it enabled by default.  To use kernel polling,
    your Linux kernel may need to be 2.6.38 or newer, and udev may need to
    be version 173 or newer.

    TO DETERMINE IF KERNEL POLLING IS ENABLED:

        cat /sys/module/block/parameters/events_dfl_poll_msecs
        cat /sys/block/sr0/events_poll_msecs

    If you get 0 or -1 from both of those commands, kernel polling may be
    disabled.

    TO ENABLE KERNEL POLLING PERMANENTLY (survives a reboot), add the
    following command to your /etc/rc.local file (anywhere before the
    'exit' line in that file):

        echo 2000 > /sys/module/block/parameters/events_dfl_poll_msecs

    Any number between 2000 and 5000 (milliseconds) should be reasonable -
    the higher 5000 means poll every 5 seconds, which is less overhead but
    a little slower.

    OR pass this option to the kernel boot command line in grub:

        block.events_dfl_poll_msecs=2000

    OR add a udev rule to enable kernel polling on removable devices:

        echo 'ACTION=="add", ATTR{removable}=="1", \
              ATTR{events_poll_msecs}=="-1", ATTR{events_poll_msecs}="2000"' \
              > /etc/udev/rules.d/61-removable-storage-polling.rules

    A reboot will be required for the above changes to take effect, or...

    TO ENABLE KERNEL POLLING TEMPORARILY AND IMMEDIATELY, enable common
    polling for the block module:

        sudo bash -c 'echo 2000 > /sys/module/block/parameters/events_dfl_poll_msecs'

    OR you can enable polling just for a single device like this (/dev/sr0
    in this example):

        sudo bash -c 'echo 2000 > /sys/block/sr0/events_poll_msecs'

    This change should be immediate - media will be detected.  However, the
    above change will be lost when you reboot.

    References:
    http://www.mail-archive.com/lfs-dev@linuxfromscratch.org/msg15714.html
    http://blogs.gentoo.org/mgorny/2011/06/20/uam-can-now-mount-cds-and-dvds/
    https://bugs.archlinux.org/task/25609
    http://unix.stackexchange.com/questions/38582/


UNINSTALL

    If you installed from a package, use your package manager to remove udevil.
    Otherwise, extract the tarball and enter the directory containing
    configure:

        # first run configure with your original options if needed:
        ./configure
        sudo make uninstall
        # Note: /etc/udevil/ is not automatically removed


TRANSLATE

    To help translate udevil please visit:
        https://github.com/IgnorantGuru/spacefm/wiki/Translate
    
    At Transifex:
        https://www.transifex.com/projects/p/udevil/

    Just sign up for a free Transifex account.  For instructions visit:
        http://help.transifex.com/intro/translating.html#translating


LICENSE

    udevil - Copyright (C) 2015   GPL3+

    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 3 of the License, or (at your
    option) any later version.

    USE AT YOUR OWN RISK.  This program is distributed in the hope that it
    will be useful, but WITHOUT ANY WARRANTY; without even the implied
    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
    the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not,  write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

    See COPYING file.



ChangeLog
0.4.4+
    include sys/stat.h in device-info.h for gcc 5.x #59
    add --always-exec option for exec during startup/--mount-all #48
0.4.4   2015-05-04:
    udevil.conf new default/allowed options for exfat #52
    [devmon 1.1.8] fix no error popup on --unmount-removable #42
    [devmon 1.1.8] improve --exec-on-unmount for mounted during startup
    [devmon 1.1.8] unset %d %l for --exec-on-unmount #56
    special handling for cifs mount ipv6 literal #46; retain literal host
    udevil.conf allowed_options_ftpfs remove ip=*; add ip= after test #46
    respect configure option --sysconfdir #31 #53
    udisks2 compatibility: replace /run/media/$USER with /media/$USER #47
    no setfacl udisks2 fallback compat user:root 0700 on /media/$USER #47
    support git-style /** suffix for some udevil.conf patterns #37
0.4.3   2013-12-09:
    fix default udevil.conf allowed_options missing fmask=0133 #35
0.4.2   2013-12-04:
    move packages to pkg branch
    debian pkgs libudev0 | libudev1
    update translations; add es fr pl sv
    fix building with automake-1.13
    autogen.sh ensure pot not deleted
    improve denied 94 msg #24
    default fmask=0133 for vfat ntfs msdos umsdos #25
    [devmon 1.1.7] support raid /dev/mda in partition detection #28
    fix identify empty optical drive
    [devmon 1.1.7] add option --enumerate-device-files #26
0.4.1   2013-03-17:
    update Russian translation
    added translation files to transifex.com
    added Italian translation by Matteo Bernardini
    added partial German translation by Delix
    add update-pot script
    clear -Werror=format-security warnings
    [devmon 1.1.6] handle hyphen in device node name
    updated German; added pt_BR Portuguese (Brazil)
0.4.0   2013-02-11:
    debian package add XXXXXXXX to mktemp template [dist]
    [devmon 1.1.5] get pmount mount point after mount #20
    move debian/ to distros/debian/ and update README
    fix -t with net fstype without protocol
    create /run on demand
    added WebDAV davfs2 support (http:// https://)
0.3.6   2012-12-16:
    file mount prevent race conditions
    disallow remount of file due to potential race conditions
    systemd service file set DISPLAY for popups #17
0.3.5   2012-11-24:
    ftpfs limit attempts on none #14
    handle mount point trailing space #12
    update russian translation
    default udevil.conf add file to allowed_types
0.3.4   2012-10-21:
    btrfs moint point detection workaround - spacefm issue 165
    adjustment for build on debian squeeze
    install devmon systemd service files (or --disable-systemd) #10
    [devmon 1.1.4] use spacefm or zenity for popup dialogs
0.3.3   2012-09-14:
    update ru.po
    no quote cifs password
    allow most UTF-8 characters in options
    default udevil.conf add utf8 to default_options_iso9660,ntfs,vfat
0.3.2   2012-08-15:
    accept smb workgroup/user
    curlftpfs pass user:pass as option
    allow email address as username
    allow unmount ejected disc #6 #7
    [devmon 1.1.3] unmount ejected disc; clean unmounted #6 #7
    [devmon 1.1.3] less verbose which
    no default ro for udf in udevil.conf
0.3.1   2012-07-25:
    cifs,smbfs use guest option when no user specified
    sshfs add port as option
    added default_options_ntfs to udevil.conf
    [devmon 1.1.2] udevil clean on remove device #6
    sshfs supplies current user if none specified
    cifs no user specified try guest mount then current user
    ready for translations - see TRANSLATE
    translation independent msg numbers in errors, warnings, and denied msgs
    test forbidden file consistently
    Russian translation by Vladimir Kudrya
0.3.0   2012-07-13:
    cifs,ftpfs dont quote username, double quote password #3
    default udevil.conf: allow option flush
    net mount point name inc dir #4
0.2.9   2012-06-29:
    fixed POTFILES.in to correct build deb problem #2
    users can mount/umount tmpfs and ramfs if in allowed_types; updated conf
    conf allow option sync for nfs; remove noatime,utf8 for smbfs/cifs #3
    change mount usage for parted magic mtab issue
0.2.8   2012-06-15:
    ssh:// support
    use cifs by default instead of smbfs
    fix denied when passing mount option uid= on command line
    updated default udevil.conf
0.2.7   2012-06-04:
    try u/mount as current user early; clean all mount points on unmount
0.2.6   2012-06-03:
    replace realpath.c with canonicalize.c
0.2.5   2012-05-30:
    correct make check error in po (affects debian build package)
    correct debian postinst error
0.2.4   2012-05-28:
    allow unmount removed devices
    [devmon 1.1.1] unmount removed devices if mounted
    mounted messages to stdout instead of stderr
    [devmon 1.1.1] use pmount ahead of udisks; test executables
    do not log read default config location
    udevil.conf: iso9660 & udf add ro to prevent mount read-only warning
0.2.3   2012-05-27:
    create /run/media/$USER on demand; add configure --with-setfacl-prog
    corrected Arch Linux udevil-git AUR package overwriting udevil.conf
0.2.2   2012-05-26:
    build with large file support for stat64 files >2GB on 32 bit fixes #1
    accept relative arguments on command line
0.2.1   2012-05-25:
    log error msgs when parsing network url
    added devmon 1.1.0 automounting script to udevil distribution
0.2.0   2012-05-22:
    initial test release



src/wildmatch.c
/*
**  Do shell-style pattern matching for ?, \, [], and * characters.
**  It is 8bit clean.
**
**  Written by Rich $alz, mirror!rs, Wed Nov 26 19:03:17 EST 1986.
**  Rich $alz is now <rsalz@bbn.com>.
**
**  Modified by Wayne Davison to special-case '/' matching, to make '**'
**  work differently than '*', and to fix the character-class code.
*/

#include "wildmatch.h"

#include <ctype.h>
#include <stddef.h>
#include <string.h>

typedef unsigned char uchar;

/* What character marks an inverted character class? */
#define NEGATE_CLASS	'!'
#define NEGATE_CLASS2	'^'

#define CC_EQ(class, len, litmatch) ((len) == sizeof (litmatch)-1 \
				    && *(class) == *(litmatch) \
				    && strncmp((char*)class, litmatch, len) == 0)

#if defined STDC_HEADERS || !defined isascii
# define ISASCII(c) 1
#else
# define ISASCII(c) isascii(c)
#endif

#ifdef isblank
# define ISBLANK(c) (ISASCII(c) && isblank(c))
#else
# define ISBLANK(c) ((c) == ' ' || (c) == '\t')
#endif

#ifdef isgraph
# define ISGRAPH(c) (ISASCII(c) && isgraph(c))
#else
# define ISGRAPH(c) (ISASCII(c) && isprint(c) && !isspace(c))
#endif

#define ISPRINT(c) (ISASCII(c) && isprint(c))
#define ISDIGIT(c) (ISASCII(c) && isdigit(c))
#define ISALNUM(c) (ISASCII(c) && isalnum(c))
#define ISALPHA(c) (ISASCII(c) && isalpha(c))
#define ISCNTRL(c) (ISASCII(c) && iscntrl(c))
#define ISLOWER(c) (ISASCII(c) && islower(c))
#define ISPUNCT(c) (ISASCII(c) && ispunct(c))
#define ISSPACE(c) (ISASCII(c) && isspace(c))
#define ISUPPER(c) (ISASCII(c) && isupper(c))
#define ISXDIGIT(c) (ISASCII(c) && isxdigit(c))

// free is_glob_special from git tree
#define GIT_GLOB_SPECIAL 0x08
const unsigned char sane_ctype[256];
#define GIT_SPACE 0x01
#define GIT_DIGIT 0x02
#define GIT_ALPHA 0x04
#define GIT_GLOB_SPECIAL 0x08
#define GIT_REGEX_SPECIAL 0x10
#define GIT_PATHSPEC_MAGIC 0x20
#define GIT_CNTRL 0x40
#define GIT_PUNCT 0x80
enum {
S = GIT_SPACE,
A = GIT_ALPHA,
D = GIT_DIGIT,
G = GIT_GLOB_SPECIAL, /* *, ?, [, \\ */
R = GIT_REGEX_SPECIAL, /* $, (, ), +, ., ^, {, | */
P = GIT_PATHSPEC_MAGIC, /* other non-alnum, except for ] and } */
X = GIT_CNTRL,
U = GIT_PUNCT,
Z = GIT_CNTRL | GIT_SPACE
};
const unsigned char sane_ctype[256] = {
X, X, X, X, X, X, X, X, X, Z, Z, X, X, Z, X, X, /* 0.. 15 */
X, X, X, X, X, X, X, X, X, X, X, X, X, X, X, X, /* 16.. 31 */
S, P, P, P, R, P, P, P, R, R, G, R, P, P, R, P, /* 32.. 47 */
D, D, D, D, D, D, D, D, D, D, P, P, P, P, P, G, /* 48.. 63 */
P, A, A, A, A, A, A, A, A, A, A, A, A, A, A, A, /* 64.. 79 */
A, A, A, A, A, A, A, A, A, A, A, G, G, U, R, P, /* 80.. 95 */
P, A, A, A, A, A, A, A, A, A, A, A, A, A, A, A, /* 96..111 */
A, A, A, A, A, A, A, A, A, A, A, R, R, U, P, X, /* 112..127 */
/* Nothing in the 128.. range */
};
#define sane_istest(x,mask) ((sane_ctype[(unsigned char)(x)] & (mask)) != 0)
#define is_glob_special(x) sane_istest(x,GIT_GLOB_SPECIAL)
/*
static int is_glob_special(char c)
{
	static const char globs[] = { '$', '(', ')', '+', '.', '^', '{', '|' };
	size_t i;
	for(i = 0; i < sizeof(globs); ++i) {
		if(globs[i] == c)
			return 1;
	}
	return 0;
}
*/

/* Match pattern "p" against "text" */
static int dowild(const uchar *p, const uchar *text, unsigned int flags)
{
	uchar p_ch;
	const uchar *pattern = p;

	for ( ; (p_ch = *p) != '\0'; text++, p++) {
		int matched, match_slash, negated;
		uchar t_ch, prev_ch;
		if ((t_ch = *text) == '\0' && p_ch != '*')
			return WM_ABORT_ALL;
		if ((flags & WM_CASEFOLD) && ISUPPER(t_ch))
			t_ch = tolower(t_ch);
		if ((flags & WM_CASEFOLD) && ISUPPER(p_ch))
			p_ch = tolower(p_ch);
		switch (p_ch) {
		case '\\':
			/* Literal match with following character.  Note that the test
			 * in "default" handles the p[1] == '\0' failure case. */
			p_ch = *++p;
			/* FALLTHROUGH */
		default:
			if (t_ch != p_ch)
				return WM_NOMATCH;
			continue;
		case '?':
			/* Match anything but '/'. */
			if ((flags & WM_PATHNAME) && t_ch == '/')
				return WM_NOMATCH;
			continue;
		case '*':
			if (*++p == '*') {
				const uchar *prev_p = p - 2;
				while (*++p == '*') {}
				if (!(flags & WM_PATHNAME))
					/* without WM_PATHNAME, '*' == '**' */
					match_slash = 1;
				else if ((prev_p < pattern || *prev_p == '/') &&
				    (*p == '\0' || *p == '/' ||
				     (p[0] == '\\' && p[1] == '/'))) {
					/*
					 * Assuming we already match 'foo/' and are at
					 * <star star slash>, just assume it matches
					 * nothing and go ahead match the rest of the
					 * pattern with the remaining string. This
					 * helps make foo/<*><*>/bar (<> because
					 * otherwise it breaks C comment syntax) match
					 * both foo/bar and foo/a/bar.
					 */
					if (p[0] == '/' &&
					    dowild(p + 1, text, flags) == WM_MATCH)
						return WM_MATCH;
					match_slash = 1;
				} else
					return WM_ABORT_MALFORMED;
			} else
				/* without WM_PATHNAME, '*' == '**' */
				match_slash = flags & WM_PATHNAME ? 0 : 1;
			if (*p == '\0') {
				/* Trailing "**" matches everything.  Trailing "*" matches
				 * only if there are no more slash characters. */
				if (!match_slash) {
					if (strchr((char*)text, '/') != NULL)
						return WM_NOMATCH;
				}
				return WM_MATCH;
			} else if (!match_slash && *p == '/') {
				/*
				 * _one_ asterisk followed by a slash
				 * with WM_PATHNAME matches the next
				 * directory
				 */
				const char *slash = strchr((char*)text, '/');
				if (!slash)
					return WM_NOMATCH;
				text = (const uchar*)slash;
				/* the slash is consumed by the top-level for loop */
				break;
			}
			while (1) {
				if (t_ch == '\0')
					break;
				/*
				 * Try to advance faster when an asterisk is
				 * followed by a literal. We know in this case
				 * that the the string before the literal
				 * must belong to "*".
				 * If match_slash is false, do not look past
				 * the first slash as it cannot belong to '*'.
				 */
				if (!is_glob_special(*p)) {
					p_ch = *p;
					if ((flags & WM_CASEFOLD) && ISUPPER(p_ch))
						p_ch = tolower(p_ch);
					while ((t_ch = *text) != '\0' &&
					       (match_slash || t_ch != '/')) {
						if ((flags & WM_CASEFOLD) && ISUPPER(t_ch))
							t_ch = tolower(t_ch);
						if (t_ch == p_ch)
							break;
						text++;
					}
					if (t_ch != p_ch)
						return WM_NOMATCH;
				}
				if ((matched = dowild(p, text, flags)) != WM_NOMATCH) {
					if (!match_slash || matched != WM_ABORT_TO_STARSTAR)
						return matched;
				} else if (!match_slash && t_ch == '/')
					return WM_ABORT_TO_STARSTAR;
				t_ch = *++text;
			}
			return WM_ABORT_ALL;
		case '[':
			p_ch = *++p;
#ifdef NEGATE_CLASS2
			if (p_ch == NEGATE_CLASS2)
				p_ch = NEGATE_CLASS;
#endif
			/* Assign literal 1/0 because of "matched" comparison. */
			negated = p_ch == NEGATE_CLASS ? 1 : 0;
			if (negated) {
				/* Inverted character class. */
				p_ch = *++p;
			}
			prev_ch = 0;
			matched = 0;
			do {
				if (!p_ch)
					return WM_ABORT_ALL;
				if (p_ch == '\\') {
					p_ch = *++p;
					if (!p_ch)
						return WM_ABORT_ALL;
					if (t_ch == p_ch)
						matched = 1;
				} else if (p_ch == '-' && prev_ch && p[1] && p[1] != ']') {
					p_ch = *++p;
					if (p_ch == '\\') {
						p_ch = *++p;
						if (!p_ch)
							return WM_ABORT_ALL;
					}
					if (t_ch <= p_ch && t_ch >= prev_ch)
						matched = 1;
					else if ((flags & WM_CASEFOLD) && ISLOWER(t_ch)) {
						uchar t_ch_upper = toupper(t_ch);
						if (t_ch_upper <= p_ch && t_ch_upper >= prev_ch)
							matched = 1;
					}
					p_ch = 0; /* This makes "prev_ch" get set to 0. */
				} else if (p_ch == '[' && p[1] == ':') {
					const uchar *s;
					int i;
					for (s = p += 2; (p_ch = *p) && p_ch != ']'; p++) {} /*SHARED ITERATOR*/
					if (!p_ch)
						return WM_ABORT_ALL;
					i = p - s - 1;
					if (i < 0 || p[-1] != ':') {
						/* Didn't find ":]", so treat like a normal set. */
						p = s - 2;
						p_ch = '[';
						if (t_ch == p_ch)
							matched = 1;
						continue;
					}
					if (CC_EQ(s,i, "alnum")) {
						if (ISALNUM(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "alpha")) {
						if (ISALPHA(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "blank")) {
						if (ISBLANK(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "cntrl")) {
						if (ISCNTRL(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "digit")) {
						if (ISDIGIT(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "graph")) {
						if (ISGRAPH(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "lower")) {
						if (ISLOWER(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "print")) {
						if (ISPRINT(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "punct")) {
						if (ISPUNCT(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "space")) {
						if (ISSPACE(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "upper")) {
						if (ISUPPER(t_ch))
							matched = 1;
						else if ((flags & WM_CASEFOLD) && ISLOWER(t_ch))
							matched = 1;
					} else if (CC_EQ(s,i, "xdigit")) {
						if (ISXDIGIT(t_ch))
							matched = 1;
					} else /* malformed [:class:] string */
						return WM_ABORT_ALL;
					p_ch = 0; /* This makes "prev_ch" get set to 0. */
				} else if (t_ch == p_ch)
					matched = 1;
			} while (prev_ch = p_ch, (p_ch = *++p) != ']');
			if (matched == negated ||
			    ((flags & WM_PATHNAME) && t_ch == '/'))
				return WM_NOMATCH;
			continue;
		}
	}

	return *text ? WM_NOMATCH : WM_MATCH;
}

/* Match the "pattern" against the "text" string. */
int wildmatch(const char *pattern, const char *text,
	      unsigned int flags, struct wildopts *wo)
{
	return dowild((const uchar*)pattern, (const uchar*)text, flags);
}
